backend/regalloc/regalloc_split_state.ml

```diff
@@ -106,6 +107,139 @@ end = struct
+                               (Cfg_dominators.is_strictly_dominating
+                                  doms.dominators spilled_at tree.label)
+                          then fatal "inconsistent dominator tree";
+                        if split_debug
+                        then
+                          log ~indent:3 "case/3 (already spilled at %d)"
+                            spilled_at;
+                        false)
+                  in
+                  if keep
+                  then
+                    already_spilled
+                      := Reg.Map.add reg tree.label !already_spilled;
+                  if split_debug then log ~indent:3 "keep? %B" keep;
+                  keep)
+                destroyed
+            in
+            if Reg.Set.is_empty destroyed
+            then (
+              if split_debug then log ~indent:3 "(the destroyed set is empty)";
+              None)
+            else Some (Destruction_on_all_paths, destroyed)
+          | Some (Destruction_only_on_exceptional_path, _) as existing ->
+            if split_debug
+            then log ~indent:2 "(ignored as a half-destruction point)";
+            existing
+          | None ->
+            if split_debug then log ~indent:2 "(not a destruction point)";
+            None)
+        destructions_at_end
+    in
+    List.fold_left tree.children ~init:destructions_at_end
+      ~f:(fun destructions_at_end (child : Cfg_dominators.dominator_tree) ->
+        if split_debug then log ~indent:2 "child %d" child.label;
+        remove_dominated_spills doms loops child ~num_sets
+          ~already_spilled:!already_spilled ~destructions_at_end)
+
+  let optimize cfg_with_liveness doms ~destructions_at_end =
+    if split_debug
+    then log ~indent:0 "RemoveDominatedSpillsForConstants.optimize";
+    let loops =
+      (* CR-soon xclerc for xclerc: be sure to not duplicate this computation if
+         for instance used to compute spilling costs. *)
+      Cfg_loop_infos.build (Cfg_with_liveness.cfg cfg_with_liveness) doms
+    in
+    let incr_set (tbl : set Reg.Tbl.t) (arr : Reg.t array) ~(in_loop : bool) :
+        unit =
+      Array.iter arr ~f:(fun (reg : Reg.t) ->
+          match Reg.Tbl.find_opt tbl reg with
+          | None ->
+            Reg.Tbl.replace tbl reg
+              (if in_loop then At_least_twice else Exactly_once)
+          | Some Exactly_once -> Reg.Tbl.replace tbl reg At_least_twice
+          | Some At_least_twice -> ())
+    in
+    let num_sets =
+      Cfg_with_liveness.fold_blocks cfg_with_liveness ~init:(Reg.Tbl.create 123)
+        ~f:(fun label block acc ->
+          let in_loop : bool = is_in_loop loops label in
+          if split_debug then log ~indent:1 "block %d in_loop? %B" label in_loop;
+          incr_set acc block.terminator.res ~in_loop;
+          DLL.iter block.body ~f:(fun (instr : Cfg.basic Cfg.instruction) ->
+              incr_set acc instr.res ~in_loop);
+          acc)
+    in
```

Comment:
This can be computed using a simple forward analysis on the CFG without `loops` whose construction is expensive and its fields are not used here, as far as I can tell (they aren't used to compute dominators, are they).

Reply:
See reply above; in short loop infos
is used because an occurrence in
`res` when in a loop is actually a
"many" set.

Reply:
(Which does not mean a forward
dataflow analysis is not possible.)

Reply:
I understand why `loops` is needed here, but it seems wasteful. forward analysis would be better.

Reply:
I am not sure: it is quite likely we will
compute the loop infos anyway eventually,
to have better spilling costs.

Reply:
ok, on second thought, this will be fine in terms of overhead after the CR in L207 is addresses and assuming that cfg_infos is computed anyway. so I'm happy with the PR to go ahead as is, although I still think it would be clearer to express this computation as forward analysis.